### PR TITLE
Add deprecation warnings for Java 7

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

As stated in [https://github.com/gradle/gradle-private/issues/935](https://github.com/gradle/gradle-private/issues/935), we should add deprecation warning for Gradle running on Java 7.

### Status

- [x] Warning on running gradle and gradlew from the command-line using Java 7, including when the build is configured to use Java 8.
- [x] Warning on using TestKit from Java 7.
- [x] Warning on using the tooling API from Java 7, including when the build is configured to use Java 8.
- [x] Test coverage that the user receives the warning in each case, and also doesn't see the warning when not using Java 7.
- [x] Make sure a user only sees at most one warning in a paticular build.

Tests passed on my local machine with both Java 7 and Java 8.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/launcher/src/main/java/org/gradle/launcher/cli/CommandLineActionFactory.java [hunk 1]
```diff
@@ -135,6 +136,9 @@ public CommandLineParseFailureAction(CommandLineParser parser, Exception e) {
         }
 
         public void execute(ExecutionListener executionListener) {
+            if (!JavaVersion.current().isJava8Compatible()) {
+                System.out.println("Support for running Gradle using Java 7 has been deprecated and is scheduled to be removed in Gradle 5.0");
+            }
```
Comment (jjohannes) [inline subprojects/launcher/src/main/java/org/gradle/launcher/cli/CommandLineActionFactory.java]: :o: Unlike `BuildExceptionReporter` this probably can not produce a duplicated message. I would still remove this to not blow up the code with warning everywhere. I think it is fine not to issue the warning is Gradle is started in a broken way.

We could keep it, but then we need some test coverage for this as well.

File: subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.gradle.launcher
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.util.Requires
+
+@Requires(adhoc = { AvailableJavaHomes.getJdks("1.7") })
+class DeprecatedBuildIntegrationTest extends AbstractIntegrationSpec {
```
Comment (jjohannes) [inline subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy]: :x: Would be good to also have a test to check that running with Java 8 does not print the warning.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/ConnectorServices.java [hunk 3]
```diff
@@ -57,6 +58,9 @@ public static void reset() {
 
     private static void checkJavaVersion() {
         UnsupportedJavaRuntimeException.assertUsingVersion("Gradle Tooling API", JavaVersion.VERSION_1_7);
+        if (!JavaVersion.current().isJava8Compatible()) {
+            DeprecationLogger.nagUserWith("Support for running Gradle using Java 7 has been deprecated and is scheduled to be removed in Gradle 5.0");
+        }
```
Comment (jjohannes) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/ConnectorServices.java]: :o: Maybe we can extend `assertUsingVersion()` such that it keeps care of calling `nagUserWith` instead of adding the individual calls to `nagUserWith`? Something like `assertUsingVersion("Gradle Tooling API", JavaVersion.VERSION_1_7, JavaVersion.VERSION_1_8)`, where we fail if VERSION_1_7 is not fulfilled and warn if VERSION_1_8 is not fulfilled.

File: subprojects/launcher/src/main/java/org/gradle/launcher/exec/InProcessBuildActionExecuter.java [hunk 4]
```diff
@@ -42,6 +46,9 @@ public Object execute(BuildAction action, BuildRequestContext buildRequestContex
             RootBuildLifecycleListener buildLifecycleListener = contextServices.get(ListenerManager.class).getBroadcaster(RootBuildLifecycleListener.class);
             buildLifecycleListener.afterStart();
             try {
+                if (!JavaVersion.current().isJava8Compatible() && (actionParameters.getLogLevel() != QUIET && actionParameters.getLogLevel() != ERROR)) {
```
Comment (bmuschko) [inline subprojects/launcher/src/main/java/org/gradle/launcher/exec/InProcessBuildActionExecuter.java]: :x: I believe this code shouldn't exist in this place. The `InProcessBuildActionExecuter` represents the mechanic to execute Gradle builds as part of an integration test. The deprecation message should come from Gradle itself and not the testing infrastructure.

File: subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerCaptureOutputIntegrationTest.groovy [hunk 5]
```diff
@@ -57,8 +59,8 @@ class GradleRunnerCaptureOutputIntegrationTest extends BaseGradleRunnerIntegrati
 
         // isn't empty if version < 2.8 or potentially contains Gradle dist download progress output
         if (isCompatibleVersion('2.8') && !crossVersion) {
-            assert stdStreams.stdOut.empty
-            assert stdStreams.stdErr.empty
+            assert normalize(stdStreams.stdOut).empty
+            assert normalize(stdStreams.stdErr).empty
```
Comment (bmuschko) [inline subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerCaptureOutputIntegrationTest.groovy]: :o: Why is this change required?

File: subprojects/test-kit/src/main/java/org/gradle/testkit/runner/internal/DefaultGradleRunner.java [hunk 6]
```diff
@@ -60,6 +69,10 @@ public DefaultGradleRunner() {
         this.gradleExecutor = gradleExecutor;
         this.testKitDirProvider = testKitDirProvider;
         this.debug = Boolean.getBoolean(DEBUG_SYS_PROP);
+
+        if (!JavaVersion.current().isJava8Compatible()) {
```
Comment (bmuschko) [inline subprojects/test-kit/src/main/java/org/gradle/testkit/runner/internal/DefaultGradleRunner.java]: :x: TestKit uses the Tooling API under the covers. The deprecations is probably not needed here as the check is already part of `DefaultConnection.java`.

File: subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerDeprecatedBuildJvmIntegrationTest.groovy [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.testkit.runner.fixtures.NoDebug
+import org.gradle.testkit.runner.fixtures.NonCrossVersion
+import spock.lang.IgnoreIf
+
+@NonCrossVersion
+class GradleRunnerDeprecatedBuildJvmIntegrationTest extends BaseGradleRunnerIntegrationTest {
+
+    @IgnoreIf({ AvailableJavaHomes.jdk7 == null })
+    @NoDebug
```
Comment (bmuschko) [inline subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerDeprecatedBuildJvmIntegrationTest.groovy]: :x: It would make sense to also test this with TestKit's debug mode.

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedBuildJvmCrossVersionSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+import org.gradle.integtests.tooling.fixture.ToolingApiSpecification
+import org.gradle.integtests.tooling.fixture.ToolingApiVersion
+import org.gradle.tooling.ProjectConnection
+import org.gradle.tooling.TestExecutionException
+import org.gradle.tooling.model.GradleProject
+import spock.lang.IgnoreIf
+
+@IgnoreIf({ AvailableJavaHomes.jdk7 == null })
+@TargetGradleVersion("current")
+class ToolingApiDeprecatedBuildJvmCrossVersionSpec extends ToolingApiSpecification {
```
Comment (bmuschko) [inline subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedBuildJvmCrossVersionSpec.groovy]: :x: Should also demonstrate that the deprecation message is not rendered with Java >= 8.

File: subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiClasspathProvider.groovy [hunk 9]
```diff
@@ -54,6 +54,7 @@ trait ToolingApiClasspathProvider {
         sharedSpec.allowPackage('junit.framework')
         sharedSpec.allowPackage('groovy')
         sharedSpec.allowPackage('org.codehaus.groovy')
+        sharedSpec.allowPackage('org.apache.commons')
```
Comment (bmuschko) [inline subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiClasspathProvider.groovy]: :x: I think it would be good to avoid allowing more external packages. The slimmer we can keep the Tooling API the better. It's really just for the use of a single method: `StringUtils.countMatches`.

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+sourceCompatibility = 1.7
+targetCompatibility = 1.7
+repositories {
+    maven {
+        url 'https://repo.gradle.org/gradle/libs-releases-local'
+    }
+    maven {
+        url '${buildContext.libsRepo.toURI()}'
+    }
+    mavenCentral()
```
Comment (bmuschko) [inline subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy]: :x: Should use `mavenCentralRepository()` from parent class.

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+    String runScript() {
+        def outStr = new ByteArrayOutputStream()
+        def executer = new ScriptExecuter()
+        executer.environment(JAVA_HOME: AvailableJavaHomes.jdk7.javaHome)
+        executer.workingDir(projectDir)
+        executer.errorOutput = outStr // simple slf4j writes warnings to stderr
+        executer.commandLine("build/install/test/bin/test")
+        executer.run().assertNormalExitValue()
+        println outStr
```
Comment (bmuschko) [inline subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy]: :x: Not needed for test.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 12]
```diff
@@ -1061,19 +1086,28 @@ private void validate(String output, String displayName) {
                     if (line.matches(".*use(s)? or override(s)? a deprecated API\\.")) {
                         // A javac warning, ignore
                         i++;
+                    } else if (line.contains("Support for running Gradle using Java 7 has been deprecated and is scheduled to be removed in Gradle 5.0")) {
```
Comment (bmuschko) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: Can you explain why we need those changes? Is this to prevent tests to fail because they are now emitting deprecation warnings if run on Java 7-based CI jobs?

File: subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy (hunk 2 above)
Comment (bmuschko) [inline subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy]: :x: This test likely needs `executer.expectDeprecationWarning()` to avoid it to fail.

**Comment by blindpirate** [comment]
@jjohannes You're right. I have already put the deprecation warning in `UnsupportedJavaRuntimeVersion.assertUsingVersion` and remove the scattered code everywhere. Every time the version is checked, the deprecation warning would be checked "by the way".

> - [x] Add test coverage checking that the warning is not printed twice in certain situation (e.g. if the build fails)
> - [x] Add test coverage that the warning is not printed when running Java8
> - [x] Put the warning text into a constant that is reused

File: subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerCaptureOutputIntegrationTest.groovy (hunk 5 above)
Comment (blindpirate) [reply subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerCaptureOutputIntegrationTest.groovy]: This test will fail on Java 7 because stdout contains a deprecation warning. 

For the tests which are `out.contains('something')`, everything gonna be okay, but for the tests which are `out=='something'` or `out=~/some exp/`, they would fail because the extra deprecation warning.

File: subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerDeprecatedBuildJvmIntegrationTest.groovy (hunk 7 above)
Comment (blindpirate) [reply subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerDeprecatedBuildJvmIntegrationTest.groovy]: Sure.

File: subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiClasspathProvider.groovy (hunk 9 above)
Comment (blindpirate) [reply subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiClasspathProvider.groovy]: Isn't this just in tests? Will it affect Tooling API jars?

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/ConnectorServices.java (hunk 3 above)
Comment (blindpirate) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/ConnectorServices.java]: Inspired by you, I put the `nagUserWith` in `checkJavaVersion` method so that each time `checkJavaVersion` is called, the java 7 deprecation warning would be called by the way.

File: subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java [hunk 13]
```diff
@@ -65,6 +66,9 @@ public void buildFinished(BuildResult result) {
     }
 
     public void execute(Throwable failure) {
+        if (!JavaVersion.current().isJava8Compatible()) {
+            System.out.println("Support for running Gradle using Java 7 has been deprecated and is scheduled to be removed in Gradle 5.0");
+        }
```
[the remark this answers is not in the document]
Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java]: This has been removed. A unified `DeprecationLogger` has been used in `UnsupportedJavaRuntimeVersion`.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 12 above)
Comment (blindpirate) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: Yes. All integration tests would be affected on jdk7. There's a deprecation warning check in this class.

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy (hunk 11 above)
Comment (blindpirate) [reply subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy]: I don't understand, could you explain it? This test exists when java 6 support was deprecated in `2.14-rc`.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/GradleConnector.java [hunk 14]
```diff
@@ -69,7 +69,7 @@
  *
  * <h2>Java version compatibility</h2>
  *
- * <p>The Tooling API requires Java 7 or later. The Gradle version used by builds may have additional Java version requirements.</p>
+ * <p>The Tooling API requires Java 7 or later. Support for Java 7 has been deprecated and is scheduled to be removed in Gradle 5.0. The Gradle version used by builds may have additional Java version requirements.</p>
```
[the remark this answers is not in the document]
Comment (blindpirate) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/GradleConnector.java]: Done.

File: subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy (hunk 2 above)
Comment (blindpirate) [reply subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy]: Done.

File: subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.launcher
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.util.Requires
+
+@Requires(adhoc = { AvailableJavaHomes.getJdks("1.7") })
```
[the remark this answers is not in the document]
Comment (blindpirate) [reply subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy]: Now some changes happened, it's `AvailableJavaHomes.getJdks("1.7", "1.8")`, please review it again, thanks!

File: subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy (hunk 2 above)
Comment (blindpirate) [reply subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy]: Well, in the beginning I was going to use `executer.expectDeprecationWarning()` - soon I found it will make things complicated and difficult to move on. Since java 7 deprecation warning is a warning which will affect all tests on JVM (and daemon, tooling api, etc.), I couldn't find a perfect approach to solve the problem. After several hours' struggling, I decided to use simple `String.contains()` instead of `executer.expectDeprecationWarning()`.

File: subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerDeprecatedBuildJvmIntegrationTest.groovy (hunk 7 above)
Comment (blindpirate) [reply subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerDeprecatedBuildJvmIntegrationTest.groovy]: Done.

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy (hunk 10 above)
Comment (blindpirate) [reply subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy]: Done.

File: subprojects/launcher/src/main/java/org/gradle/launcher/exec/InProcessBuildActionExecuter.java (hunk 4 above)
Comment (blindpirate) [reply subprojects/launcher/src/main/java/org/gradle/launcher/exec/InProcessBuildActionExecuter.java]: I agree with you, but if I remove the deprecation warning here, [this test](https://github.com/gradle/gradle/blob/3512ecab86759b8e203b96a609cb607b98c7f0f0/subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedBuildJvmCrossVersionSpec.groovy#L61) will fail on jdk7. My understanding is: if a jdk 7 JVM (`client`) connects to a jdk 7 JVM (`build`) and do some build, the warning should appear in the build's output. However, in this case, `client` JVM's deprecation warning has already appeared during its startup, so it wouldn't appear again (`DeprecationLogger`), and `build` JVM has no chance to run the version checking code. That's why it need to be here. 

Please let me know if I didn't make myself clear or I didn't understand the code correctly.

**Comment by blindpirate** [comment]
@melix FYI

File: subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java [hunk 16]
```diff
@@ -18,15 +18,26 @@
 
 import org.gradle.api.GradleException;
 import org.gradle.api.JavaVersion;
+import org.gradle.util.DeprecationLogger;
 import org.gradle.util.GradleVersion;
 
 public class UnsupportedJavaRuntimeException extends GradleException {
+    public static final String JAVA7_DEPRECATION_WARNING = "Support for running Gradle using Java 7 has been deprecated and is scheduled to be removed in Gradle 5.0. "
```
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: :o: It probably will probably be long until we need to deprecate another Java version, but I think it would be nice to have this more general. So that next time deprecating a Java versions would only require us to change one constant. I am thinking of something like:

```
private static final JavaVersion RECOMMENDED_VERSION = JavaVersion.VERSION_1_8;
    
public static final String JAVA_DEPRECATION_WARNING = "Support for running Gradle using Java" + RECOMMENDED_VERSION.getMajorVersion() +" has been deprecated and is scheduled to be removed in Gradle 5.0. "
        + "Please see https://docs.gradle.org/current/userguide/java_plugin.html#sec:java_cross_compilation for more details.";

    private static void javaDeprecationWarning() {
        if (JavaVersion.current().compareTo(RECOMMENDED_VERSION) >= 0) {
            DeprecationLogger.nagUserWith(JAVA_DEPRECATION_WARNING);
        }
    }
```

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy (hunk 11 above)
Comment (melix) [reply subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy]: I think that Ben talks about the `println` statement.

File: subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiClasspathProvider.groovy (hunk 9 above)
Comment (melix) [reply subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiClasspathProvider.groovy]: No, it won't affect the TAPI jars, you're right @blindpirate . However I agree that it would be better to avoid the use of `StringUtils`, as we would expose something we don't need in general, so there's a risk of introducing unnoticed regressions in the future.

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy (hunk 11 above)
Comment (blindpirate) [reply subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy]: Oh, of course, thanks! @melix @bmuschko

File: subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java [hunk 17]
```diff
@@ -18,15 +18,26 @@
...
+        + "Please see https://docs.gradle.org/current/userguide/java_plugin.html#sec:java_cross_compilation for more details.";
```
Comment (melix) [inline subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: Maybe we can improve the error message a bit here, and be more explicit. Instead of "hey, RTFM", which is what the message says today, we can say:

> Support for running Gradle using Java 7 has been deprecated. It is however still possible to compile, test an run applications that use Java 7 using cross-compilation. See .... for more details.

File: subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiClasspathProvider.groovy (hunk 9 above)
Comment (blindpirate) [reply subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiClasspathProvider.groovy]: Looks like Groovy has a [count](http://docs.groovy-lang.org/docs/groovy-1.7.3/html/groovy-jdk/java/lang/String.html#count(java.lang.String)), I should use that instead.

File: subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java (hunk 17 above)
Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: Well... is it too long for a console output?

Comment (melix) [reply subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: I'm not sure. I've always been thinking most error messages in Gradle are too terse. @jjohannes @bmuschko WDYT?

Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: Should we use `DocumentationRegistry` to get the right URL to the userguide?

Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: Yes, you're right. I will do that.

Comment (melix) [reply subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: Good catch!

File: subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java (hunk 16 above)
Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: Good catch and good advice, thanks!

Comment (blindpirate) [reply subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: I think I've fix this, let me wait and see the CI results.

File: subprojects/launcher/src/main/java/org/gradle/launcher/exec/InProcessBuildActionExecuter.java (hunk 4 above)
Comment (blindpirate) [reply subprojects/launcher/src/main/java/org/gradle/launcher/exec/InProcessBuildActionExecuter.java]: I don't think so... it looks like all build operations will pass this code (and I'm inspired by [here](https://github.com/gradle/gradle/blob/ec877ed3bfc83e6905985da13cf0507dd8f05af4/subprojects/launcher/src/main/java/org/gradle/launcher/exec/InProcessBuildActionExecuter.java#L44)).

After struggling with failed tests, I found that deprecation warnings should be here so that bootstrap JVM and build JVM wouldn't generate duplicate warnings.

**Comment by bmuschko** [comment]
@blindpirate Can you please change the base branch of this PR to release now that we branched for the 4.2 release?

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy (hunk 11 above)
Comment (bmuschko) [reply subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiDeprecatedClientJvmCrossVersionSpec.groovy]: Yeah, I just mean the `println`.

File: subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java (hunk 17 above)
Comment (bmuschko) [reply subprojects/core/src/main/java/org/gradle/internal/jvm/UnsupportedJavaRuntimeException.java]: I think it's OK to be more explicit here. In addition we will also want to add a link that allows a user to give feedback whether the deprecation would be a blocker for them. @eriwen Is there a way you could provide us with a URL that redirects to a specific page as mentioned by @bigdaz [here](https://github.com/gradle/gradle-private/issues/935#issuecomment-324648508)?

File: subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.internal.jvm.UnsupportedJavaRuntimeException
+import org.gradle.util.Requires
+import spock.lang.IgnoreIf
+
+@Requires(adhoc = { AvailableJavaHomes.getJdks("1.7", "1.8") })
+class DeprecatedBuildIntegrationTest extends AbstractIntegrationSpec {
+    @IgnoreIf({ AvailableJavaHomes.jdk7 == null })
+    def "warns of deprecated java version when running under java 7"() {
+        given:
```
Comment (bmuschko) [inline subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy]: :o: I feel like that the tests would read better if you'd use `given`, `when`, `then` e.g.

```
given:
executer.withJavaHome(AvailableJavaHomes.jdk7.javaHome)

when:
run("help")

then:
warningCount() == 1
```

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 19]
```diff
@@ -992,6 +992,14 @@ private boolean isJava7Home(String path){
...
     }
 
+    private boolean isJava7DeprecationTruncatedDaemonLog(List<String> lines, int i, String line) {
+        // for forkingIntegrationTests running on Java 7, daemon exit log will fail the tests with following stdout:
+        // ----- Last  20 lines from daemon log file - daemon-6753.out.log -----
+        //    at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
+        //    at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
+        //    ...
+        return line.startsWith("----- Last  ") && i < lines.size() - 1 && STACK_TRACE_ELEMENT.matcher(lines.get(i + 1)).matches();
+    }
```
Comment (jjohannes) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: :x: This does not look right. The daemon should not exit with an exception. Can you point me at the test(s) that are/were failing?

Comment (blindpirate) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: This has nothing to do with the daemon disappear issue. This is to fix this case:

On Java 7, in forkingIntegTest, deprecation warning will be printed into daemon log, then truncated and displayed on forkingIntegTest's stdout, which breaks tests with 'unexpected stacktrace'. This code is just a workaround to avoid modifying dozens of forkingIntegTests and should be removed in subsequent development.

File: subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy (hunk 18 above)
Comment (blindpirate) [reply subprojects/launcher/src/integTest/groovy/org/gradle/launcher/DeprecatedBuildIntegrationTest.groovy]: Good advice!

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 19 above)
Comment (jjohannes) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: Alright. I got that wrong. Thanks for clarifying.
